python/k4FWCore/ApplicationMgr.py

```diff
@@ -37,6 +40,67 @@ class ApplicationMgr:
     def __init__(self, **kwargs):
         self._mgr = AppMgr(**kwargs)
 
+    def _setup_reader(self, reader, iosvc_props):
+        """Setup the reader consistently such that it has sane defaults
+
+        In some cases we have to peek into the files to obtain some information
+        to set sane default values if they are not set by the user. We need to
+        obtain
+        - The number of events in case EvtMax is set to -1
+        - The collection names in case the CollectionNames are not provided
+
+        Knowing the number of events is necessary to avoid errors when running
+        multithreaded since if we have, for example, 10 events and we are
+        running 9 at the same time, then (possibly) the first 9 complete and 9
+        more are scheduled, out of which only one will be finished without
+        errors. If we know the number of events in advance then we can just
+        schedule those.
+
+        We need the collection names to read to feed them to the Reader. Either
+        we take the user provided ones or we get them from the first event in
+        the file we peek into.
+        """
+        # First we determine whether we need to peek at all
+        inp = None
+        if iosvc_props["input"][0]:
+            inp = "input"
+        elif iosvc_props["Input"][0]:
+            inp = "Input"
+
+        if not inp:
+            # We have got nothing to do here, since there is no input
+            return
+
+        collections = iosvc_props["CollectionNames"][0] or None
+        n_events = self._mgr.EvtMax
+
+        # Check if we can get by without peeking into the file
+        if collections:
+            logger.info(f"Setting the reader to read the collections: {collections}")
+            reader.InputCollections = collections
+            if n_events != -1:
+                # We know everything we need
+                logger.info(f"Initializing reader to read {n_events} events")
+                return
+
+        # We need to peek into the file because we lack information.
+        # Import here to avoid always importing ROOT which is slow
+        from podio.root_io import Reader as PodioReader
+
+        podio_reader = PodioReader(iosvc_props[inp][0])
+        if n_events == -1:
+            self._mgr.EvtMax = len(podio_reader.get("events"))
+        if not collections:
+            try:
+                frame = podio_reader.get("events")[0]
+                logger.debug("Using the first frame to determine collections to read")
+                collections = list(frame.getAvailableCollections())
+            except IndexError:
+                print("Warning, the events category wasn't found in the input file")
```

Comment:
```suggestion
                logger.warning("Warning, the events category wasn't found in the input file")
```
Since we use the logging module in other places.